Dapper.Tests.Database/Tests/TestSuiteUpdate.cs

```diff
@@ -187,5 +188,32 @@ public void UpdatePartial()
                 Assert.Equal("Smith", gp.LastName);
             }
         }
+
+        [Fact]
+        [Trait("Category", "Update")]
+        public void UpdateConcurrencyCheck()

```

Comment:
We're testing for the failed use case with the `[ConcurrencyCheck]` attribute, but not the success where it doesn't fail on a field marked with `[ConcurrencyCheck]`, correct?

For all the tests testing update, we should make sure that it properly allows for an update when the attribute is used in the WHERE query generator and the row was *not* updated in another connection context.

I wrote a quick test copy of this that did just that, and it succeeds, but would be good to add them for all updates for the explicit behavior we want.

Additionally, there's no limits on specifying multiple `[ConcurrencyCheck]` columns in your model from what I can see, so we should consider adding a model with two properties using that attribute to ensure that we don't need to enforce only specifying it once on a model. (As I understand the way you built/wrote this, this shouldn't be an issue, it should just work, but we should consider adding that test case since we're not enforcing only having the attribute specified once on the model)

Reply:
The updates have a single test for concurrency success and failure, but I'll go ahead and split the tests to make it clearer.

The main reason I avoided doing multiple columns was to avoid changing the tables as much as possible, but I'll go ahead and do that so the tests cover that.

Reply:
OK, concurrency checks tests have been rewritten to have separate modified/not-modified tests for Update, UpdateList, Upsert sync/async variants.

While doing this I uncovered a bug wherein we weren't properly comparing nullable ConcurrencyCheck fields, as the SQL generated did not handle null semantics.  So I added metadata to ColumnInfo to tell if a field was nullable.

The following determines whether a field is nullable or not:

* Keys (those marked with [KeyAttribute]) are not null.
* Non-keys marked with [RequiredAttribute] are not null.
* Otherwise, nullness is determined by data type:
    * Value types are not nullable unless they are of type `Nullable<T>`.
    * For reference types, the new C# 8.0 [nullable references](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references) are honored in a similar fashion when nullable checks are enabled (more formally, when they are within a nullable annotation context).
    * When nullable checks are not enabled (C# 7.3 or earlier, or `#nullable disable`), reference types are considered nullable.

[KeyAttribute]: https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.keyattribute
[RequiredAttribute]: https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.requiredattribute